src/default_storage.jl

```diff
@@ -0,0 +1,393 @@
+    return storage.inputs_map
+end
+
+
+function Salsa._previous_output_internal(
+    runtime::Salsa._TracingRuntimeWithStorage{DefaultStorage},
+    key::DependencyKey{<:DerivedKey},
+)
+    storage = _storage(runtime)
+    derived_key, args = key.key, key.args
+
+    previous_output = nothing
+
+    try
+        lock(storage.lock)
+        cache = get_map_for_key(storage, derived_key)
+
+        if haskey(cache, args)
+            previous_output = getindex(cache, args)
+        end
+    finally
+        unlock(storage.lock)
+    end
+
+    return previous_output
+end
+
+
+# TODO: I think we can @nospecialize the arguments for compiler performance?
+# TODO: It doesn't seem like this @nospecialize is working... It still seems to be compiling
+# a nospecialization for every argument type. :(
+function Salsa._memoized_lookup_internal(
+    runtime::Salsa._TracingRuntimeWithStorage{DefaultStorage},
+    key::DependencyKey{<:DerivedKey},
+)
+    storage = _storage(runtime)
+    try  # For storage.derived_functions_active
+        try
+            lock(storage.lock)
+            storage.derived_functions_active += 1
+        finally
+            unlock(storage.lock)
+        end
+
+        existing_value = nothing
+        value = nothing
+
+        derived_key, args = key.key, key.args
+
+        # NOTE: We currently make no attempts to prevent two Tasks from simultaneously
+        # computing the same derived function for the same key. For cheap derived functions
+        # this may be a more optimal behavior than the overhead caused by coordination.
+        # However for expensive functions, this is not ideal. We take this approach for now
+        # because it is simpler.
+
+        # The locking here is a bit involved, to prevent the lock from being held during
+        # user computation.
+        # We minimally lock only around the accesses to the fields of storage.
+        # NOTE: It is okay to release and then reacquire the lock inside this function
+        # while computing the result because we are guaranteed that inputs cannot be
+        # modified while this function is running, so we do not need to fear concurrency
+        # violations, e.g. overwriting newer values with outdated results.
+        lock_held::Bool = false
+        local cache
+        try
+            lock(storage.lock)
+            lock_held = true
+
+            cache = get_map_for_key(storage, derived_key)
+
+            if haskey(cache, args)
+                # TODO: Optimization idea:
+                #   - There's no reason to be tracing the Salsa functions during
+                #     the `still_valid` check, since we're not going to use them. We might
+                #     _do_ still want to keep the stack trace though for cycle detection and
+                #     error messages.
+                #   - We might want to consider keeping some toggle on the Trace object
+                #     itself, to allow us to skip recording the deps for this phase.
```

Comment:
Opened https://github.com/RelationalAI-oss/Salsa.jl/issues/20 to track this TODO